File: Difficulty: Medium/Directed Graph Cycle/directed-graph-cycle.cpp
#define graph_t vector<vector<int>>
#define visit_t vector<bool>
#define vi vector<int>
class Solution {
  public:
    bool isCyclic(int V, vector<vector<int>> &edges) {
        // code here
        graph_t adj(V);
        vi indeg(V,0),ans;
        for(auto& vec: edges) {
            adj[vec[0]].push_back(vec[1]);
            indeg[vec[1]]++;
        }
        queue<int> que;
        
        for(int i  = 0;i<V;i++) {
            if(indeg[i]==0) {
                que.push(i);
            }
        }
        
        while(!que.empty()) {
            int u = que.front();
            que.pop();
            ans.push_back(u);
            for(auto& v: adj[u]) {
                indeg[v]--;
                if(indeg[v]==0) {
                    que.push(v);
                }
            }
        }
        // cout<<ans.size()<<endl;
        if(ans.size()!=V) return true;
        return false;
    }
};

File: Difficulty: Medium/Directed Graph Cycle/directed-graph-cycle.java
class Solution {
    public boolean isCyclic(int V, int[][] edges) {
        // code here
        int n = edges.length;
        ArrayList<ArrayList<Integer>> adjlist = new ArrayList<>();
        int[] indeg = new int[V];
        // boolean[] visited = new boolean[V];
        
        for(int i = 0;i<V;i++) adjlist.add(new ArrayList<Integer>());
        
        
        
        for(int i = 0;i<n;i++) {
            int u = edges[i][0];
            int v = edges[i][1];
            adjlist.get(u).add(v);
            indeg[v]++;
        }
        
        Queue<Integer> que = new ArrayDeque<Integer>();
        
        for(int i = 0;i<V;i++) {
            if(indeg[i]==0) {
                que.add(i);
                // visited[i]=true;
            }
        }
        
        int ans = 0;
        
        while(!que.isEmpty()) {
            int u = que.poll();
            ans++;
            for(Integer v: adjlist.get(u)) {
                indeg[v]--;
                if(indeg[v]==0) {
                    que.add(v);
                    // visited[v]=true;
                }
            }
        }
        
        if(ans==V) return false;
        return true;
    }
}